lib/web/kube.go

```diff
@@ -0,0 +1,344 @@
+	}
+
+	stream := NewTerminalStream(ctx, TerminalStreamConfig{WS: p.ws, Logger: p.log})
+
+	certsReq := clientproto.UserCertsRequest{
+		PublicKey:         userKey.MarshalSSHPublicKey(),
+		Username:          p.sctx.GetUser(),
+		Expires:           p.sctx.cfg.Session.GetExpiryTime(),
+		Format:            constants.CertificateFormatStandard,
+		RouteToCluster:    p.teleportCluster,
+		KubernetesCluster: p.req.KubeCluster,
+		Usage:             clientproto.UserCertsRequest_Kubernetes,
+	}
+
+	_, certs, err := client.PerformMFACeremony(ctx, client.PerformMFACeremonyParams{
+		CurrentAuthClient: p.userClient,
+		RootAuthClient:    p.sctx.cfg.RootClient,
+		MFAPrompt: mfa.PromptFunc(func(ctx context.Context, chal *clientproto.MFAAuthenticateChallenge) (*clientproto.MFAAuthenticateResponse, error) {
+			assertion, err := promptMFAChallenge(stream.WSStream, protobufMFACodec{}).Run(ctx, chal)
+			return assertion, trace.Wrap(err)
+		}),
+		MFAAgainstRoot: p.sctx.cfg.RootClusterName == p.teleportCluster,
+		MFARequiredReq: &clientproto.IsMFARequiredRequest{
+			Target: &clientproto.IsMFARequiredRequest_KubernetesCluster{KubernetesCluster: p.req.KubeCluster},
+		},
+		ChallengeExtensions: mfav1.ChallengeExtensions{
+			Scope: mfav1.ChallengeScope_CHALLENGE_SCOPE_USER_SESSION,
+		},
+		CertsReq: &certsReq,
+		Key:      userKey,
+	})
+	if err != nil && !errors.Is(err, services.ErrSessionMFANotRequired) {
+		return trace.Wrap(err, "failed performing mfa ceremony")
+	}
+
+	if certs == nil {
+		certs, err = p.sctx.cfg.RootClient.GenerateUserCerts(ctx, certsReq)
+		if err != nil {
+			return trace.Wrap(err, "failed issuing user certs")
+		}
+	}
+
+	rsaKey, err := userKey.PrivateKey.RSAPrivateKeyPEM()
+	if err != nil {
+		return trace.Wrap(err, "failed getting rsa private key")
+	}
+
+	restConfig, err := createKubeRestConfig(p.configServerAddr, p.configTLSServerName, hostCA, certs.TLS, rsaKey)
+	if err != nil {
+		return trace.Wrap(err, "failed creating Kubernetes rest config")
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return trace.Wrap(err, "failed creating Kubernetes client")
+	}
+
+	kubeReq := kubeClient.CoreV1().RESTClient().Post().Resource("pods").Name(p.req.Pod).
+		Namespace(p.req.Namespace).SubResource("exec")
+	option := &v1.PodExecOptions{
+		Container: p.req.Container,
+		Command:   strings.Split(p.req.Command, " "),
+		TTY:       p.req.IsInteractive,
+		Stdin:     p.req.IsInteractive,
+		Stdout:    true,
+		Stderr:    true,
+	}
+
+	kubeReq.VersionedParams(option, scheme.ParameterCodec)
+	p.log.Debugf("Web kube exec request URL: %s", kubeReq.URL())
+
+	wsExec, err := remotecommand.NewWebSocketExecutor(restConfig, "POST", kubeReq.URL().String())
+	if err != nil {
+		return trace.Wrap(err, "failed creating websocket executor")
+	}
+
+	if err := wsExec.StreamWithContext(ctx, remotecommand.StreamOptions{
+		Stdin:  stream,
+		Stdout: stream,
+		Stderr: stderrWriter{stream: stream},
```

Comment:
stderr can only be properly used in kube when interactive is not set.

Reply:
As I understand it doesn't matter if we set it here, as long as we correctly set it on `PodExecOptions`?

Reply:
Only if aligned otherwise everything will hang forever. URL args give information about the number of Spdy streams to create and both parties must agree. In the websocket case, since streams are fake, it can probably appear to work but only for websocket

Reply:
OK, I've changed it to also not be set in case of interactive: https://github.com/gravitational/teleport/pull/41144/commits/23b67d885c314a153966dd635f54541955486767